container.go

```diff
@@ -215,34 +214,42 @@ func (c *Container) registerObject(o interface{}, otype reflect.Type) error {
 }
 
 // constr must be a function that returns the result type and an error
-func (c *Container) registerConstructor(constr interface{}) error {
+func (c *Container) registerConstructor(constr interface{}, count int) error {
 	ctype := reflect.TypeOf(constr)
-	objType := ctype.Out(0)
+	objTypes := make([]reflect.Type, count, count)
+	for i := 0; i < count; i++ {
+		objTypes[i] = ctype.Out(i)
+	}
 
+	nodes := make([]node, count, count)
+	for i := 0; i < count; i++ {
+		nodes[i] = node{
+			objType: objTypes[i],
+		}
+	}
 	argc := ctype.NumIn()
 	n := funcNode{
 		deps:        make([]interface{}, argc),
 		constructor: constr,
-		node: node{
-			objType: objType,
-		},
+		nodes:       nodes,
 	}
 	for i := 0; i < argc; i++ {
 		arg := ctype.In(i)
 		if arg.Kind() != reflect.Ptr && arg.Kind() != reflect.Interface {
 			return errArgKind
 		}
-
 		n.deps[i] = arg
 	}
 
-	c.nodes[objType] = &n
+	for i := 0; i < count; i++ {
+		c.nodes[objTypes[i]] = &n
+	}
 
 	// object needs to be part of the container to properly detect cycles
 	if cycleErr := c.detectCycles(&n); cycleErr != nil {
 		// if the cycle was detected delete from the container
-		delete(c.nodes, objType)
-		return errors.Wrapf(cycleErr, "unable to Provide %v", objType)
+		delete(c.nodes, objTypes[0])
+		return errors.Wrapf(cycleErr, "unable to Provide %v", objTypes[0])
```

Comment:
the error should be better here. iterate over the nodes and construct one string that joins all the types failed to provide, i.e. `failed to provide types A, B, C`